ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/state/OutputBuffer.java

```diff
@@ -358,26 +358,26 @@ public final synchronized void packString (String s) throws IOException
         if (!_valid)
             throw new IOException(tsLogger.i18NLogger.get_state_OutputBuffer_10());
 
-        int sz = 0;
-        String dummy = null;
-
-        if (s != null)
-        {
-            sz = s.length() + 1;
-            dummy = s + '\0';
+        if(s == null) {
+            packInt(0);
+        } else {
+            packStringBytes( s.getBytes(StandardCharsets.UTF_8) );
+        }
+    }
 
+    public final synchronized void packStringBytes(byte[] bytes) throws IOException
+    {
+        if (!_valid || bytes == null) {
+            throw new IOException(tsLogger.i18NLogger.get_state_OutputBuffer_10());
         }
 
-        packInt(sz);
+        packInt(bytes.length+1);
 
         _valid = false;
 
-        if (sz > 0)
-        {
-            byte[] bytes = dummy.getBytes(StandardCharsets.UTF_8);
-            _output.write(bytes, 0, bytes.length);
-            realign(bytes.length);
-        }
+        _output.write(bytes, 0, bytes.length);
+        _output.writeByte(0);
```

Comment:
Just a question - did the original code add space for the string delimiter but not actually pack it into the output buffer? How did you uncover this bug?

Reply:
The original works by creating a new string with the terminator appended as a charater, then sizing/encoding that in a single step, rather than treating the string and the terminator as separate elements. The latter is more efficient, as it avoids creating an additional String object.